### ISSUE TITLE
cljs.loader does not work due to incorrect module URIs generated by boot-cljs

### ISSUE THREAD

**Issue by esp1** [boot-clj/boot-cljs#174]
cljs-boot does not work with the new ClojureScript code splitting functionality's `cljs.loader` as described here: https://clojurescript.org/guides/code-splitting

I am using the code exactly as shown in the preceding link, but to build I am using the following `resources/hello.cljs.edn` file:
```
{:compiler-options
  {:optimizations :advanced
   :pseudo-names true
   :modules {:foo {:entries #{foo.core}
                   :output-to "out/foo.js"}
             :bar {:entries #{bar.core}
                   :output-to "out/bar.js"}}}}
```
and this `build.boot` file:
```
(set-env!
  :resource-paths #{"resources"}
  :source-paths #{"src"}
  :dependencies '[[adzerk/boot-cljs "2.1.3" :scope "test"]
                  [org.clojure/clojurescript "1.9.908"]])

(require '[adzerk.boot-cljs :refer [cljs]])

(deftask build []
  (comp (cljs)
        (target)))
```

When I load the web page and attempt to load the `bar` module by clicking on the button, I get this error in the browser console:
```
cljs_base.js:1342 GET http://localhost:3000/hello.out/Users/esp/.boot/cache/tmp/Users/esp/Code/tmp/hello-modules/11sw/-pphjkn/hello/bar.js 404 (Not Found)
```

Note how the URL to bar.js is misconstructed with the absolute path to the file on the filesystem embedded within it.

Printing out the value of `cljs.loader/module-uris` also shows the malformed module URIs:
```
MODULE_URIS 
{:cljs-base [hello.out/cljs_base.js],
 :foo [hello.out/Users/esp/.boot/cache/tmp/Users/esp/Code/tmp/hello-modules/11w7/-pphjkn/out/foo.js],
 :bar [hello.out/Users/esp/.boot/cache/tmp/Users/esp/Code/tmp/hello-modules/11w7/-pphjkn/out/bar.js]}
```
Note that the only value that is correct is the one for `:cljs-base`, which does not specify an `:output-to` compiler option. Removing the `:output-to` specification for the modules however still results in badly formed URIs:
```
MODULE_URIS
{:cljs-base [hello.out/cljs_base.js],
 :foo [hello.out/Users/esp/.boot/cache/tmp/Users/esp/Code/tmp/hello-modules/11sw/-pphjkn/hello/foo.js],
 :bar [hello.out/Users/esp/.boot/cache/tmp/Users/esp/Code/tmp/hello-modules/11sw/-pphjkn/hello/bar.js]}
```

**Comment by Deraen** [comment]
cljs-base path is not correct because that points to working directory, and not the temp-dirs.

All the output paths should point to the temp-dirs.

Not sure why loading is broken.

**Comment by esp1** [comment]
No, the `:cljs-base` **URI** is correct. This is not the same as the filesystem path the compiler outputs to. The MODULE_URIS are what are used by the `cljs.loader` to load module js files via network requests.

The URIs for the `:foo` and `:bar` modules are incorrect. Note how they begin with the relative path to the asset dir, and then are followed by the `:output-to` value as an absolute path. E.g.:

**hello.out**/Users/esp/.boot/cache/tmp/Users/esp/Code/tmp/hello-modules/11sw/-pphjkn/**hello/foo.js**

The ClojureScript compiler automatically constructs these module URIs from the compiler options. I'm speculating that when it does this the compiler is implicitly assuming that all the paths in the module compiler options are relative, and the absolute paths to the temp directory are messing it up.

**Comment by Deraen** [comment]
Boot-cljs doesn't set the URI's, only output paths. Not sure how these module uris are created.

**Comment by esp1** [comment]
I've been playing around with this outside of `boot-cljs` using different values of `:output-dir`, `:asset-path`, and `:output-to`.

The values specified in the Code Splitting guide work, where `:output-dir`, `:asset-path`, and the prefix dir for all of the `:output-to` values are all `out`:
```
{:output-dir "out"
 :asset-path "/out"
 :modules {:foo {:entries #{foo.core}, :output-to "out/foo.js"}
           :bar {:entries #{bar.core}, :output-to "out/bar.js"}}}
```
```
MODULE_URIS
{:cljs-base [out/cljs_base.js],
 :foo [out/foo.js],
 :bar [out/bar.js]}
```
This also works:
```
{:output-dir "resources/public/out"
 :asset-path "out"
 :modules {:foo {:entries #{foo.core}, :output-to "resources/public/out/foo.js"}
           :bar {:entries #{bar.core}, :output-to "resources/public/out/bar.js"}}}
```
```
MODULE_URIS
{:cljs-base [out/cljs_base.js],
 :foo [out/foo.js],
 :bar [out/bar.js]}
```
But changing `:output-dir` to something that doesn't match the prefix of the `:output-to` paths specified in the `:modules` configurations results in malformed paths:
```
{:output-dir "hello.out"
 :asset-path "out"
 :modules {:cljs-base {:output-to "out/cljs_base.js"}
           :foo {:entries #{foo.core}, :output-to "out/foo.js"}
           :bar {:entries #{bar.core}, :output-to "out/bar.js"}}}
```
```
MODULE_URIS
{:cljs-base [out/Users/esp/Code/tmp/hello-modules/out/cljs_base.js],
 :foo [out/Users/esp/Code/tmp/hello-modules/out/foo.js],
 :bar [out/Users/esp/Code/tmp/hello-modules/out/bar.js]}
```
Note the malformed URIs start with the `:asset-path` and are followed by the absolute path of the module `:output-to` path. Examining the malformed URIs from the `boot-cljs` example above:
```
hello.out/Users/esp/.boot/cache/tmp/Users/esp/Code/tmp/hello-modules/11sw/-pphjkn/out/foo.js
```
`boot-cljs` has assigned `:output-dir` and `:asset-path` to `hello.out`, but that doesn't match the prefix of the `:output-to` path I specified for the `:foo` module (`out/foo.js`). If I change the `:output-to` paths in the `hello.cljs.edn` to e.g `hello.out/foo.js`, it works.

So now I know what's going on, and how to fix it. Question though: should `boot-cljs` perhaps automatically prefix the `:output-dir` path onto the module `:output-to` paths so you don't need to manually do this yourself when using modules?

**Comment by esp1** [comment]
Also, if the `:output-to` is omitted from the module configurations, `boot-cljs` automatically creates one using `hello` (the name of the .cljs.edn file, minus the extension) as a prefix instead of `hello.out` (the value of `:output-dir`), which also causes the malformed URI issue. Minimally I think it would be a good idea to change this behavior to use the `:output-dir` value to prefix automatically generated module `:output-to` values.

**Comment by Deraen** [comment]
So I got this working with one configuration:

```
{:compiler-options
  {:optimizations :advanced
   :pseudo-names true
   :modules {:foo {:entries #{foo.core}
                   :output-to "hello.out/foo.js"}
             :bar {:entries #{bar.core}
                   :output-to "hello.out/bar.js"}}}}
```

Letting boot-cljs set output-to automatically based module name doesn't seem to work, as that uses `hello` instead of `hello.out` as dir. Will fix that.

The example where you use `:output-to "out/foo.js"` is broken, I think, if you don't also provide `:output-dir "out"` option:

```
{:compiler-options
  {:optimizations :advanced
   :pseudo-names true
   :output-dir "out"
   :modules {:foo {:entries #{foo.core}
                   :output-to "out/foo.js"}
             :bar {:entries #{bar.core}
                   :output-to "out/bar.js"}}}}
```

(Letting boot-cljs set output-to automatically doesn't work in this case either)

**Comment by Deraen** [comment]
@esp1 https://github.com/boot-clj/boot-cljs/commit/9d23492928108d8ec0c62fc9876c11c095c6512c I believe this will fix problems when letting boot-cljs set `:output-to` based on module name.

**Comment by esp1** [comment]
Great, thanks!

**Comment by Deraen** [comment]
I deployed 2.1.4-SNAPSHOT, please test that. I'll release proper version after I find a reason or fix for #172, if  it doesn't take too long.

**Comment by esp1** [comment]
Works for me - thanks again!
